### DISCUSSION TITLE
Implementation into Unity

### DISCUSSION THREAD

**Discussion by rombodawg** [oobabooga/text-generation-webui#796] [General]
Im trying to find a way to implement webui into unity for creation of advanced npc's running on local hardware without the need for cloud or subscription based models like open-ai gpt-3/gpt-4. Im having trouble figuring out a few things though as I am a novice in both the ai and game development space, but I hope to create something great for the gaming community with this already awesome open sourced tool. My current hurdles are figuring out how the web ui installer generates the unique url for each users webpage to use the program. Ill need to know this in order to be able to hook into it using unity webhook feature. 

Also I if anyone knows if the web-ui is actually open source and free for commercial use and would like to chime in that would be great? If the developer oobabooga sees this id really appreciate a response because id love to do a colab to create a unity asset to be able to hook into your webui to for game creation. Even if you dont reveal the source code to me, but just tell me how parts of it works so i can create a unity web hook based around it.

**Comment by dexterhuh** [comment]
Hi Rombodawg, You cannot implement this into Unity. They are different farmwork. If you want your NPCs that use a language model you don't need to implement this into Unity, just implement it somewhere and use Unity to communicate with it. I never tried this, but here is my thought.

1. Set and implement it onto any computer as a server, this server could be a local server, so any computer that can run it and can be reached really.
2. Use Python to make a communicator to communicate serving Vicuna or other models from 1 via the API (search API in the discussions for example https://github.com/oobabooga/text-generation-webui/discussions/1191#discussioncomment-5639754).
3. Make Unity communicate with your Python communicator to give text from your player as input to the server and receive the output from the server that displays in your NPC dialog (refer to this https://github.com/Siliconifier/Python-Unity-Socket-Communication)

Let me know if you have any problems, and let's discuss them.

**Comment by BetaDoggo** [comment]
The suggestion above is definitely the best and intended way to go about this.

About licensing and commercial use:
This repo is protected by the AGPLv3 license. This essentially means that the source code can be modified for commercial purposes but any modifications to it must be provided to end users. This only extends to the webui itself so if you build a unity asset to communicate with the unmodified api as suggested you should be fine.

Another Note:
Each model has its own license that you have to comply with as well so some models may not be usable for commercial purposes. As long as you don't bundle any models with the asset you shouldn't have to worry about this though.
